source/auth/tests/mongodb-oidc.rst

```diff
@@ -4,146 +4,437 @@ MongoDB OIDC
+  minute, and a refresh callback that returns data not conforming to
+  the ``OIDCRequestTokenResult`` with extra field(s).
+- Create a client with the callbacks.
+- Perform a ``find`` operation that succeeds.
+- Close the client.
+- Create a new client with the same callbacks.
+- Perform a ``find`` operation that fails.
+- Close the client.
 
-AWS Device Auth
-===============
+Cached Credentials
+==================
 
-Drivers MUST be able to authenticate using the "aws" device workflow simulating
-an EC2 instance with an enabled web identity token provider, generated by
-Drivers Evergreen Tools.
+Drivers MUST ensure that they are testing the ability to cache credentials.
+Unless otherwise specified, the tests MUST be performed with the callback-driven workflow with a provided request and refresh callback. If
+desired, the caching tests MAY be done using mock server responses.
+The following tests assume a global cache is in use.  If a different
+cache scheme is in use, appropriate tests MUST be written to ensure that
+the cache is performing as intended.
+
+Cache with refresh
+~~~~~~~~~~~~~~~~~~
+- Clear the cache.
+- Create a new client with a request callback that gives credentials that
+  expire in on minute.
+- Ensure that a ``find`` operation adds credentials to the cache.
+- Close the client.
+- Create a new client with the same request callback and a refresh callback.
+- Ensure that a ``find`` operation results in a call to the refresh callback.
+- Close the client.
+
+Cache with no refresh
+~~~~~~~~~~~~~~~~~~~~~
+- Clear the cache.
+- Create a new client with a request callback that gives credentials that
+  expire in one minute.
+- Ensure that a ``find`` operation adds credentials to the cache.
+- Close the client.
+- Create a new client with the a request callback but no refresh callback.
+- Ensure that a ``find`` operation results in a call to the request callback.
+- Close the client.
+
+Cache key includes callback
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+If the driver does not support using callback references or hashes as part of
+the cache key, skip this test.  This test ensures that the callback is
+considered as part of the cache key.
+
+- Clear the cache.
+- Create a new client with a request callback that does not give an
+  ```expiresInSeconds``` value.
+- Ensure that a ``find`` operation adds credentials to the cache.
+- Close the client.
+- Create a new client with a different request callback.
+- Ensure that a ``find`` operation replaces the one-time use entry and adds a new entry to the cache.
+- Close the client.
+
+Error clears cache
+~~~~~~~~~~~~~~~~~~
+- Clear the cache.
+- Create a new client with a valid request callback that gives credentials
+  that expire within 5 minutes and a refresh callback that gives invalid
+  credentials.
+- Ensure that a ``find`` operation adds a new entry to the cache.
+- Ensure that a subsequent ``find`` operation results in an error.
+- Ensure that the cached access token has been cleared.
```

Comment:
Following up on [earlier](https://github.com/mongodb/specifications/pull/1381#discussion_r1163048854). I believe the spec now says that Access Tokens are cleared only when the operation fails due to a 391. It seems that we should have a test that confirms that Access Tokens are preserved when a find fails for a reason that is not 391 (like division by 0 in the query).

Reply:
The full cache is cleared during an error.  I updated the test.

Reply:
I don't think we should hold up this PR for more prose tests.

Reply:
Alright, added https://jira.mongodb.org/browse/DRIVERS-2616 .

We discussed, and only 391 errors should trigger a cache clear (the aforementioned division by zero should not). Seems fine to clarify this as part of 2616.